### ISSUE TITLE
{IndexError}index 2048 is out of bounds for dimension 1 with size 2048

### ISSUE THREAD

**Issue by zfhxi** [mxl1990/tsne-pytorch#3]
Hello, my test code is as follow:

```
X1 = torch.randn([10, 2048]).cuda()
X_emb = TorchTSNE(n_components=2, perplexity=100., n_iter=1000).fit_transform(X1)
```

Sometimes, it will occur '{IndexError}index 2048 is out of bounds for dimension 1 with size 2048', I find the reason is that as  121 line in `the tsne_pytorch.py`, there may be out of bounds:
```
119 for i in range(d):
120     if l[i, 1] != 0:
121         M[:, i+1] = M[:, i]
122         i += 1
```
Concretely, `i` may equal to d-1 in 121 line which causes M[:,i+1] will be out of bounds(we know M is dxd).

I wonder if the modified code as follow is right?
```
for i in range(d-1):
    if l[i, 1] != 0:
        M[:, i+1] = M[:, i]
        i += 1
```

**Comment by mxl1990** [comment]
> Hello, my test code is as follow:
> 
> ```
> X1 = torch.randn([10, 2048]).cuda()
> X_emb = TorchTSNE(n_components=2, perplexity=100., n_iter=1000).fit_transform(X1)
> ```
> 
> Sometimes, it will occur '{IndexError}index 2048 is out of bounds for dimension 1 with size 2048', I find the reason is that as 121 line in `the tsne_pytorch.py`, there may be out of bounds:
> 
> ```
> 119 for i in range(d):
> 120     if l[i, 1] != 0:
> 121         M[:, i+1] = M[:, i]
> 122         i += 1
> ```
> 
> Concretely, `i` may equal to d-1 in 121 line which causes M[:,i+1] will be out of bounds(we know M is dxd).
> 
> I wonder if the modified code as follow is right?
> 
> ```
> for i in range(d-1):
>     if l[i, 1] != 0:
>         M[:, i+1] = M[:, i]
>         i += 1
> ```

rang(d) function do not cover the index d, and d is get from (n, d) = X.shape
so may be something else wrong not here

**Comment by zfhxi** [comment]
> > Hello, my test code is as follow:
> > ```
> > X1 = torch.randn([10, 2048]).cuda()
> > X_emb = TorchTSNE(n_components=2, perplexity=100., n_iter=1000).fit_transform(X1)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Sometimes, it will occur '{IndexError}index 2048 is out of bounds for dimension 1 with size 2048', I find the reason is that as 121 line in `the tsne_pytorch.py`, there may be out of bounds:
> > ```
> > 119 for i in range(d):
> > 120     if l[i, 1] != 0:
> > 121         M[:, i+1] = M[:, i]
> > 122         i += 1
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Concretely, `i` may equal to d-1 in 121 line which causes M[:,i+1] will be out of bounds(we know M is dxd).
> > I wonder if the modified code as follow is right?
> > ```
> > for i in range(d-1):
> >     if l[i, 1] != 0:
> >         M[:, i+1] = M[:, i]
> >         i += 1
> > ```
> 
> rang(d) function do not cover the index d, and d is get from (n, d) = X.shape so may be something else wrong not here

You are right, but i may equal to d-1, then M[:,i+1] actually is M[:,d] which will occur errors.

**Comment by mxl1990** [comment]
> > > Hello, my test code is as follow:
> > > ```
> > > X1 = torch.randn([10, 2048]).cuda()
> > > X_emb = TorchTSNE(n_components=2, perplexity=100., n_iter=1000).fit_transform(X1)
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Sometimes, it will occur '{IndexError}index 2048 is out of bounds for dimension 1 with size 2048', I find the reason is that as 121 line in `the tsne_pytorch.py`, there may be out of bounds:
> > > ```
> > > 119 for i in range(d):
> > > 120     if l[i, 1] != 0:
> > > 121         M[:, i+1] = M[:, i]
> > > 122         i += 1
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Concretely, `i` may equal to d-1 in 121 line which causes M[:,i+1] will be out of bounds(we know M is dxd).
> > > I wonder if the modified code as follow is right?
> > > ```
> > > for i in range(d-1):
> > >     if l[i, 1] != 0:
> > >         M[:, i+1] = M[:, i]
> > >         i += 1
> > > ```
> > 
> > 
> > rang(d) function do not cover the index d, and d is get from (n, d) = X.shape so may be something else wrong not here
> 
> You are right, but i may equal to d-1, then M[:,i+1] actually is M[:,d] which will occur errors.

I got it. The matrix M is a dxd matrix which contains n eigenvectors in X.t() * X matrix
I changed the iteration function to avoid this error, thank you for your reminder
